// convnetjs-ts/lib/index.js
Object.defineProperty(exports, "__esModule", { value: true });
// direct export
var convnet_vol_1 = require("./convnet_vol");
exports.Vol = convnet_vol_1.Vol;
var convnet_net_1 = require("./convnet_net");
exports.Net = convnet_net_1.Net;
var convnet_magicnet_1 = require("./convnet_magicnet");
exports.MagicNet = convnet_magicnet_1.MagicNet;
var convnet_util_1 = require("./convnet_util");
exports.randf = convnet_util_1.randf;
exports.randi = convnet_util_1.randi;
exports.randn = convnet_util_1.randn;
exports.randperm = convnet_util_1.randperm;
var convnet_layers_dotproducts_1 = require("./convnet_layers_dotproducts");
exports.ConvLayer = convnet_layers_dotproducts_1.ConvLayer;
exports.FullyConnLayer = convnet_layers_dotproducts_1.FullyConnLayer;
var convnet_layers_dropout_1 = require("./convnet_layers_dropout");
exports.DropoutLayer = convnet_layers_dropout_1.DropoutLayer;
var convnet_layers_loss_1 = require("./convnet_layers_loss");
exports.RegressionLayer = convnet_layers_loss_1.RegressionLayer;
exports.SVMLayer = convnet_layers_loss_1.SVMLayer;
exports.SoftmaxLayer = convnet_layers_loss_1.SoftmaxLayer;
var convnet_layers_normalization_1 = require("./convnet_layers_normalization");
exports.LocalResponseNormalizationLayer = convnet_layers_normalization_1.LocalResponseNormalizationLayer;
var convnet_layers_pool_1 = require("./convnet_layers_pool");
exports.PoolLayer = convnet_layers_pool_1.PoolLayer;
// module export
var cnnvis = require("./cnnvis");
exports.cnnvis = cnnvis;
var cnnutil = require("./cnnutil");
exports.cnnutil = cnnutil;
var util = require("./convnet_util");
exports.util = util;
var volutil = require("./convnet_vol_util");
exports.volutil = volutil;
var deepqlearn = require("./deepqlearn");
exports.deepqlearn = deepqlearn;
// rename
var convnet_trainers_1 = require("./convnet_trainers");
exports.Trainer = convnet_trainers_1.Trainer;
exports.SGDTrainer = convnet_trainers_1.Trainer;
